File: flink-app/src/main/java/com/example/ctr/infrastructure/flink/sink/ClickHouseSink.java
package com.example.ctr.infrastructure.flink.sink;

import com.example.ctr.config.ClickHouseProperties;
import com.example.ctr.domain.model.CTRResult;
import org.apache.flink.connector.jdbc.JdbcConnectionOptions;
import org.apache.flink.connector.jdbc.JdbcExecutionOptions;
import org.apache.flink.connector.jdbc.JdbcSink;
import org.apache.flink.connector.jdbc.JdbcStatementBuilder;
import org.apache.flink.streaming.api.functions.sink.SinkFunction;

import java.sql.PreparedStatement;
import java.sql.SQLException;

public class ClickHouseSink {

    private final ClickHouseProperties clickHouseProperties;

    public ClickHouseSink(ClickHouseProperties clickHouseProperties) {
        this.clickHouseProperties = clickHouseProperties;
    }

    public SinkFunction<CTRResult> createSink() {
        return JdbcSink.sink(
                "INSERT INTO ctr_results (product_id, ctr, impressions, clicks, window_start, window_end) VALUES (?, ?, ?, ?, ?, ?)",
                (JdbcStatementBuilder<CTRResult>) ClickHouseSink::bindCtrResult,
                JdbcExecutionOptions.builder()
                        .withBatchSize(1000)
                        .withBatchIntervalMs(200)
                        .withMaxRetries(3)
                        .build(),
                new JdbcConnectionOptions.JdbcConnectionOptionsBuilder()
                        .withUrl(clickHouseProperties.getUrl())
                        .withDriverName(clickHouseProperties.getDriver())
                        .build());
    }

    private static final int IDX_PRODUCT_ID = 1;
    private static final int IDX_CTR = 2;
    private static final int IDX_IMPRESSIONS = 3;
    private static final int IDX_CLICKS = 4;
    private static final int IDX_WINDOW_START = 5;
    private static final int IDX_WINDOW_END = 6;

    private static void bindCtrResult(PreparedStatement ps, CTRResult ctrResult) throws SQLException {
        ps.setString(IDX_PRODUCT_ID, ctrResult.getProductId());
        ps.setDouble(IDX_CTR, ctrResult.getCtr());
        ps.setLong(IDX_IMPRESSIONS, ctrResult.getImpressions());
        ps.setLong(IDX_CLICKS, ctrResult.getClicks());
        ps.setLong(IDX_WINDOW_START, ctrResult.getWindowStart());
        ps.setLong(IDX_WINDOW_END, ctrResult.getWindowEnd());
    }
}

File: flink-app/src/main/kotlin/com/example/ctr/infrastructure/flink/sink/ClickHouseSink.kt
package com.example.ctr.infrastructure.flink.sink

import com.example.ctr.config.ClickHouseProperties
import com.example.ctr.domain.model.CTRResult
import org.apache.flink.connector.jdbc.JdbcConnectionOptions
import org.apache.flink.connector.jdbc.JdbcExecutionOptions
import org.apache.flink.connector.jdbc.JdbcSink
import org.apache.flink.connector.jdbc.JdbcStatementBuilder
import org.apache.flink.streaming.api.functions.sink.SinkFunction
import java.sql.PreparedStatement

class ClickHouseSink(private val clickHouseProperties: ClickHouseProperties) {

    fun createSink(): SinkFunction<CTRResult> = JdbcSink.sink(
        "INSERT INTO ctr_results (product_id, ctr, impressions, clicks, window_start, window_end) VALUES (?, ?, ?, ?, ?, ?)",
        JdbcStatementBuilder<CTRResult> { ps, ctrResult -> bindCtrResult(ps, ctrResult) },
        JdbcExecutionOptions.builder()
            .withBatchSize(1000)
            .withBatchIntervalMs(200)
            .withMaxRetries(3)
            .build(),
        JdbcConnectionOptions.JdbcConnectionOptionsBuilder()
            .withUrl(clickHouseProperties.url)
            .withDriverName(clickHouseProperties.driver)
            .build()
    )

    companion object {
        private const val IDX_PRODUCT_ID = 1
        private const val IDX_CTR = 2
        private const val IDX_IMPRESSIONS = 3
        private const val IDX_CLICKS = 4
        private const val IDX_WINDOW_START = 5
        private const val IDX_WINDOW_END = 6

        private fun bindCtrResult(ps: PreparedStatement, ctrResult: CTRResult) {
            ps.setString(IDX_PRODUCT_ID, ctrResult.productId)
            ps.setDouble(IDX_CTR, ctrResult.ctr)
            ps.setLong(IDX_IMPRESSIONS, ctrResult.impressions)
            ps.setLong(IDX_CLICKS, ctrResult.clicks)
            ps.setLong(IDX_WINDOW_START, ctrResult.windowStart)
            ps.setLong(IDX_WINDOW_END, ctrResult.windowEnd)
        }
    }
}
